# Acquire Exercises

In [98]:
from bs4 import BeautifulSoup
import re

import requests

# 1. Codeup Blog Articles

Visit Codeup's Blog and record the urls for at least 5 distinct blog posts. For each post, you should scrape at least the post's title and content.

Encapsulate your work in a function named get_blog_articles that will return a list of dictionaries, with each dictionary representing one article. The shape of each dictionary should look like this:

```python
{
    'title': 'the title of the article',
    'content': 'the full text content of the article'
}
```
Plus any additional properties you think might be helpful.

Bonus: Scrape the text of all the articles linked on codeup's blog page.

In [2]:
url='https://codeup.com/blog/'

In [26]:
soup = BeautifulSoup(requests.get(url).content, 'html.parser')

In [27]:
soup

<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr/><center>nginx</center>
</body>
</html>

In [23]:
html_text =requests.get(url).text

In [24]:
print(html_text)

<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr><center>nginx</center>
</body>
</html>



I got code 403; trying to figure a way to acess the information.

In [28]:
#Define headers
headers = {'User-Agent': 'Codeup Data Science'}

In [32]:
#Make a request
response = requests.get('https://codeup.com/data-science/become-a-data-scientist/', headers=headers)

In [124]:
#Create the soup and investigate
soup = BeautifulSoup(response.content, 'html.parser')


soup

In [48]:
title= soup.find('h1')
title.text

'Become a Data Scientist in 6 Months!'

In [49]:
#Access the date published
publish_date = soup.find('span', class_='published')
publish_date.text

'Jan 5, 2023'

In [50]:
#Access the article content
content = soup.find('div', class_='entry-content')
content.text

'\nAre you feeling unfulfilled in your work but want to avoid returning to the traditional educational route? Codeup can help! Starting over as a professional is daunting and not always ideal. Codeup can help you go from a career you are bored with, to a job that excites you in just 6 months!\nHere’s how…\nData Science Program\nDuring our 20-week program, you will have the opportunity to take your career to new heights with data science being one of the most needed jobs in tech.\nYou’ll gather data, then clean it, explore it for trends, and apply machine learning models to make predictions.\nUpon completing this program, you will know how to turn insights into actionable recommendations. You’ll be a huge asset to any company, having all the technical skills to become a data scientist with projects upon projects of experience under your belt.\nCodeup\nA common reason individuals opt not to change their careers is fear it is too late. Codeup has crafted a program that will guide you thro

In [52]:
article_links = soup.find_all('a', 'href')
article_links

[]

In [71]:
# Get the "a" tag
element = soup.find('a')
# Get the attribute value
data = element.get('href')
data

'https://codeup.com/'

In [77]:
# Get the "a" tag
element = soup.find_all('a')
# Get the attribute value
element


[<a href="https://codeup.com/">Home</a>,
 <a href="https://codeup.com/program/cloud-adminsitration/">Cloud Administration</a>,
 <a href="https://codeup.com/program/full-stack-web-development/">Full Stack Web Development</a>,
 <a href="https://codeup.com/program/data-science/">Data Science</a>,
 <a href="/events/">Workshops</a>,
 <a href="/san-antonio-events/">San Antonio</a>,
 <a href="/dallas-events/">Dallas</a>,
 <a href="https://codeup.com/financial-aid/">Financial Aid</a>,
 <a href="https://codeup.com/veterans/">Military</a>,
 <a href="https://codeup.com/hire-tech-talent/">Hire Tech Talent</a>,
 <a href="https://alumni.codeup.com/">Alumni</a>,
 <a href="https://codeup.com/resources/">Resources</a>,
 <a href="/my-story/">Student Reviews</a>,
 <a href="https://codeup.com/blog/">Blog</a>,
 <a href="https://codeup.com/frequently-asked-questions/">Common Questions</a>,
 <a href="https://codeup.com/podcast/">Hire Tech Podcast</a>,
 <a href="https://codeup.com/apply-now/">Apply Now</a>,
 

In [79]:
tags = soup.find_all("li", "search-result")
tags

[]

In [80]:
tag = soup.find("div", id="bar")
tag

In [83]:
tags = soup.select("#search-results .external-links")
tags

[]

In [85]:
anchor_href = soup.find("a")["href"]
anchor_href

'https://codeup.com/'

In [104]:
link_list=[]
for link in soup.find_all('a'):
    link_list.append(link.get('href'))


In [115]:
link_list=[]
for link in soup.find_all('a'):
    pattern = re.compile(r'(codeup.com\/|www.|twitter.|alumni.)[\w\/-]+|(.com|.org)[\w\/-?]+|codeup.com')
    
    matches = pattern.finditer(link.get('href'))
    print(matches)

TypeError: expected string or bytes-like object

In [119]:
blog_url_list= ['https://codeup.com/data-science/become-a-data-scientist/',
                'https://codeup.com/employers/hiring-tech-talent/',
               'https://codeup.com/cloud-administration/cap-funding-options/'
               'https://codeup.com/dallas-info/it-professionals-dallas/',
               'https://codeup.com/codeup-news/codeup-voted-1-technical-school-in-dfw/',
               'https://codeup.com/tips-for-prospective-students/financing/codeups-scholarships/']

In [138]:
def basic_info(url):
    blog_info={"title","publish_date","content"}
    headers = {'User-Agent': 'Codeup Data Science'}
    response = requests.get(url, headers=headers)
    
    #find the title
    soup = BeautifulSoup(response.content, 'html.parser')
    title = soup.find('h1')
    
    
    #Access the date published
    publish_date = soup.find('span', class_='published')
 
    
    #Access the article content
    content = soup.find('div', class_='entry-content')

    output = {
        'title': title.text,
        'publish_date': publish_date.text
        'content': content.text
     }
    return output

SyntaxError: invalid syntax (1972321223.py, line 21)

In [129]:
metrics= []
for i in blog_url_list:
    output = basic_info(i)
    metrics.append(output)
metrics

Become a Data Scientist in 6 Months!
Jan 5, 2023
Hiring Tech Talent Around the Holidays
Dec 16, 2022
Why Dallas is a Great Location for IT Professionals
Dec 1, 2022
Codeup is ranked #1 Best in DFW 2022
Nov 21, 2022
Codeup’s Scholarship Offerings
Nov 17, 2022


[None, None, None, None, None]

# 2. News Articles

We will now be scraping text data from inshorts, a website that provides a brief overview of many different topics.

Write a function that scrapes the news articles for the following topics:

Business
Sports
Technology
Entertainment
The end product of this should be a function named get_news_articles that returns a list of dictionaries, where each dictionary has this shape:
``` python

{
    'title': 'The article title',
    'content': 'The article content',
    'category': 'business' # for example
}
```
Hints:

a. Start by inspecting the website in your browser. Figure out which elements will be useful.

b. Start by creating a function that handles a single article and produces a dictionary like the one above.

c. Next create a function that will find all the articles on a single page and call the function you created in the last step for every article on the page.

d. Now create a function that will use the previous two functions to scrape the articles from all the pages that you need, and do any additional processing that needs to be done.

# 3. Bonus: cache the data

Write your code such that the acquired data is saved locally in some form or fashion. Your functions that retrieve the data should prefer to read the local data instead of having to make all the requests everytime the function is called. Include a boolean flag in the functions to allow the data to be acquired "fresh" from the actual sources (re-writing your local cache).